GOALS.md

```diff
@@ -0,0 +1,233 @@
+dozens, if not hundreds, of other plugins. Not to mention the processing the DAW itself needs to do for mixing all of
+these, and the I/O it needs to perform to the hardware to get actual sound. All of it in very tight timing budgets.
+
+For instance, audio processing code may want to spread work across multiple threads to be as fast as possible. This
+works fine in a standalone application where the process can allocate most or all of the CPU cores to itself. However,
+this can't work in a session where there are many instances of that code competing for CPU power at the same time. The
+OS would have to interrupt and reschedule threads constantly, losing most of the CPU power in context switches.
+
+In general, the LV2 APIs consider the host to be in charge of handling most of the plugin's resources, and behave
+like a scheduler or executor of sorts for the various plugins. Because it knows the state of all the plugins running,
+as well as their (potentially complex) I/O configuration, it can apply massive optimizations to its scheduling and
+processing, such as spreading the work on a single thread pool. However, this means the plugins must cooperate, and
+cannot do what they want. Examples of work LV2 plugins should defer to the host are I/O buffers and communication, state
+serialization (i.e., presets / session loading and saving), asynchronous processing, or UI communication.
+
+In that aspect, LV2 plugins and hosts share very similar behaviors with Rust's own
+[Futures](https://doc.rust-lang.org/std/future/trait.Future.html) and
+[Executors](https://docs.rs/futures/0.3.16/futures/executor/index.html).
+
+Like with Rust Futures, the execution system for LV2 plugins is inherently *cooperative*. LV2 plugins *must* finish
+processing before another plugin can run. If a plugin, just like a Future, uses blocking operations, it will block
+the whole thread without any means of interruption by the host. This includes (but is not limited to): memory
+allocations, multi-thread processing, thread synchronization (atomics, locks), general I/O, and more.
+
+If a plugin needs to perform asynchronous work for instance (like loading and decoding a sample file), they should use
+the [LV2 Worker](http://lv2plug.in/ns/ext/worker) API (see the `lv2-worker` crate). Just like a Future would use the
+executor's [`spawn()`](https://docs.rs/futures/0.3.16/futures/task/trait.SpawnExt.html#method.spawn) method to process
+additional asynchronous work, instead of doing it synchronously or spawning a thread of its own.
+
+## Extra ergonomics and sugar
+Also, while ergonomics and extra utilities are nice to have in this library sometimes, they *must* be optional to use.
+Indeed, because of the goal to be a low-level library, we must not prevent the user from doing custom things themselves
+at the cost of complexity. At least, as long as it is safe for them to do so.
```

Comment:
A safe abstraction usually introduces some ergonomics, so I wouldn't say that this must be optional. 

I think designing safe abstractions will very often come with hard choices: a more complicated API, some runtime overhead (see previous section for that), something that is hard to maintain or exposing unsafe methods or traits.

Reply:
I agree, what I meant here was ergonomic*-only* APIs should be optional (like the derive macros cited in the other comment), not all APIs that happen to be more ergonomic that raw LV2's.

Pretty much all safe APIs are more ergonomic than LV2's anyway. :upside_down_face: 

I agree that API design comes with hard choices. I forgot to make it explicit in the `DESIGN` document (which isn't completed anyway), but this is what it is about: the list at the beginning is in order of most important to least important.

For example, when it comes to the `lv2` crate, Memory Safety will be chosen in favored of C Parity or Performance, all of which are considered more important than Ergonomics (which is quite low the list).

Again, will fix that. Thanks a lot for your remarks, they are quite helpful! :slightly_smiling_face: